frontend/src/app/income-source-edit/income-source-edit.component.spec.ts

```diff
@@ -0,0 +1,185 @@
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { IncomeSourceEditComponent } from './income-source-edit.component';
+import { AppModule } from '../app.module';
+import { RouterTestingModule } from '@angular/router/testing';
+import { ActivatedRoute, Router } from '@angular/router';
+import { IncomeService } from '../income.service';
+import { Observable } from 'rxjs/Observable';
+import { IncomeSourceModel } from '../models/income.model';
+
+describe('IncomeSourceEditComponent', () => {
+
+  const incomeSourceTypes = [
+    { id: 1, type: 'B' },
+    { id: 2, type: 'A' }
+  ];
+
+  describe('in creation mode', () => {
+    const activatedRoute = {
+      snapshot: { data: { incomeSourceTypes } }
+    };
+
+    beforeEach(() => TestBed.configureTestingModule({
+      imports: [AppModule, RouterTestingModule],
+      providers: [{ provide: ActivatedRoute, useValue: activatedRoute }]
+    }));
+
+    it('should expose sorted income source types', () => {
+      const fixture = TestBed.createComponent(IncomeSourceEditComponent);
+      fixture.detectChanges();
+
+      const component = fixture.componentInstance;
+      expect(component.incomeSourceTypes.map(t => t.type)).toEqual(['A', 'B']);
+    });
+
+    it('should expose a default income source', () => {
+      const fixture = TestBed.createComponent(IncomeSourceEditComponent);
+      fixture.detectChanges();
+
+      const component = fixture.componentInstance;
+      expect(component.incomeSource).toEqual({ name: '', typeId: null, maxMonthlyAmount: null });
+    });
+
+    it('should display the income source in a form, and have the save button disabled', () => {
```

Comment:
It's especially hard to know for components using ngModel
I think here async is not enough, you either have to use fakeAsync or whenStable, so that's fine